### PR TITLE
Tudumvc

### PR DESCRIPTION
**Pull request by rabsef-bicrym**
This pull request would add the %tudumvc Guide to the Urbit.org /docs/guides area. It lists itself as the last item in that list (just under aqua.md).

%tudumvc is a step by step tutorial/walkthrough of implementing [TodoMVC React.js + Hooks](https://github.com/tastejs/todomvc/tree/4e301c7014093505dcf6678c8f97a5e8dee2d250/examples/react-hooks) with an Urbit %gall agent. It uses the following progression to instruct on how to develop on Urbit appropriately:

1. Prepare a Development Environment
2. Host any file using file-server
3. Host a file using a %gall agent
4. Write a %gall agent to accommodate some data
5. Connect the front-end to the %gall agent
6. Improve front-end/%gall agent to allow for networked use through %gall

This push relates to the following [grant](https://grants.urbit.org/bounties/1372696624)

There are two images in this pr that require hosting - from looking at another open pr it looks like I need to request this then commit the changes so - consider this requested (attaching the images here).

devops

final-product

### CODE REVIEW

**Comment by jalehman** [comment]
Would love your take on this @baudtack — @timlucmiptev and myself have looked over it a good deal already and your feedback would be welcome.

File: hoon/guides/tudumvc.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
++++
```
Comment (baudtack) [inline hoon/guides/tudumvc.md:1]: this will not build with zola as hoon/guides/tudumvc.md  and hoon/guides/tudumvc/_index.md are conflicting paths.

**Review by baudtack — CHANGES_REQUESTED** [review]
I have a few minor comments but i was having difficulty giving a proper review because github is terrible and the order these things are in matters. So I tried to build it locally and it currently does not build. Please build this with the latest master urbit.org repo and let me know when it works so I can give the content a proper review.

File: hoon/guides/tudumvc.md (hunk 1 above)
Comment (rabsef-bicrym) [reply hoon/guides/tudumvc.md:1]: @baudtack Thank you for the guidance here - I had not considered building the product to test locally and was trying just to emulate style. Much appreciate the correction - working on it now.

Comment (baudtack) [reply hoon/guides/tudumvc.md:1]: @rabsef-bicrym yeah It bites me in the butt all the time when i try do to that also. If you need a hand getting it working let me know.

Comment (rabsef-bicrym) [reply hoon/guides/tudumvc.md:1]: @baudtack Is there a nesting depth limit on the sidebar? I'm having trouble getting a 4th one to show - which may not be a problem - I can reconfigure things to so accommodate.

Comment (baudtack) [reply hoon/guides/tudumvc.md:1]: I think there is but i'm not really sure about that. @matildepark do you know?

**Comment by matildepark** [comment]
I capped the sidebar iteration logic at the nesting level we had at the time.

**Comment by rabsef-bicrym** [comment]
@matildepark Assuming that I have a guide w/ an introduction and 6 "lessons" plus several "breakout lessons" that are subordinate to the  primary lessons but not required for getting the content, and that these should either go in Hoon School or Guides (probably guides), which of the following structures better represents the design aesthetic:

* Requesting an additional level of indentation
* Parsing the guide like this:
  - Lesson 1
  - Lesson 1 breakout lesson
  - Lesson 2
* Parsing the guide like this:
  - Lesson 1
  - ...
  - Lesson 6
  - Breakout Lesson 1
  - Breakout Lesson 2
* Having a hidden folder (not displayed in the indentation because of the limit) and linking to that content anyway where appropriate for the breakout lessons.

**Comment by jalehman** [comment]
I think this is more of a tutorial than a guide, and it’s about userspace development on the web rather than Hoon. In that tree I’d make it a sibling of Gall, probably immediately following it. A good title might be “Web Development Tutorial”.

—

~wolref-podlex
https://urbit.org ( https://urbit.org/ )

**Comment by rabsef-bicrym** [comment]
@baudtack It's compiling now for me so I think we're set for your review. Please let me know if it still gives issues - first time contributor, long time reader.

**Comment by jfranklin9000** [comment]
`userspace/tudumvc/agent-supported-hosting.md`:
"because it' TodoMVC" -> "because it's TodoMVC"

File: userspace/tudumvc/updating-the-agent.md [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+First, you'll need to open the `/sur/tudumvc.hoon` file and define our state as a type so that we can reference it easily in our /app file (not necessary, but idiomatic - one could define the state structures in /app, as well). Recall that we import `/sur/tudumvc.hoon` using `/-` in our /app file.
+
+#### `/sur/tudumvc.hoon`
+Our current file only defines our `action` type. For `%tudumvc` to communicate well with TodoMVC (Earth app), you'll want to add a new type called `tasks` that is structured to accommodate a `map` with a key of a task's "id" and a value for that key of the task's "label" and "done"-ness (completion state). Using this structure will allow us to index a given task by its "id" (a unique key) to modify the "label" or "done"ness (which can be non-unique - we can enter "Call Mom" 100 times and not have an indexing conflict - btw, you should call your mother - it's almost Mother's Day).
+
+Just as with `action`, we're going to use [`+$`](https://urbit.org/docs/reference/hoon-expressions/rune/lus/#lusbuc):
+<table>
+<tr>
+<td>
+:: initial version
```
Comment (baudtack) [inline userspace/tudumvc/updating-the-agent.md:64]: This is a comment for most of these two versions of code side by side thing. It's a good idea but with our current site layout it makes both sections nearly unreadable. Maybe it should be something like the tabbed option here? https://urbit.org/getting-started/

File: userspace/tudumvc/updating-the-agent.md [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+```
+
+##### `%edit-task`
+We're not going to tell you how `%edit-task` works - in fact, that's one of your exercises this part of the guide. We can tell you it works rather similarly to the ones above, and that it takes two arguments:
+* An `id`.
+* An updated `label`.
+
+## Homework {#homework}
+* Examine the available structures of [JSON in Hoon](https://github.com/urbit/urbit/blob/6bcbbf8f1a4756c195a324efcf9515b6f288f700/pkg/arvo/sys/lull.hoon#L40), found in `lull.hoon`.
+* Read through [`++  enjs`](https://github.com/urbit/urbit/blob/6bcbbf8f1a4756c195a324efcf9515b6f288f700/pkg/arvo/sys/zuse.hoon#L3263) in `zuse.hoon`.
```
Comment (baudtack) [inline userspace/tudumvc/updating-the-agent.md:528]: Generally we remove spaces in gate names so `++enjs` if you want to have the ++ at the start.

Comment (baudtack) [reply userspace/tudumvc/updating-the-agent.md:528]: This is done in a number of places in a lot of files.

**Review by baudtack — CHANGES_REQUESTED** [review]
So I have a few stylistic problems. The voice is also wildly different from everything else on our site, but I suspect that's actually fine.

File: userspace/tudumvc/updating-the-agent.md (hunk 3 above)
Comment (rabsef-bicrym) [reply userspace/tudumvc/updating-the-agent.md:528]: Oh - and as part of the revision to tutorial, I'm taking out the side by side, preferring just code blocks of the updated content per "chapter" in the tutorial.  If I find any need for side-by-side again, I will use the tabbed version as suggested above.

Comment (baudtack) [reply userspace/tudumvc/updating-the-agent.md:528]: It's just not the same as it's done on the rest of the docs is my only concern.

Comment (rabsef-bicrym) [reply userspace/tudumvc/updating-the-agent.md]: @baudtack Pushed the additional changes throughout - let me know if that style works better for the site.  Thank you again for your patience while I worked through that.

**Review by baudtack — APPROVED** [review]
Great work! This is ready to go in.

**Comment by baudtack** [comment]
@jalehman do you know of a reason this can't be merged yet?

**Comment by matildepark** [comment]
Much apologies — you'll have to retarget this against https://github.com/urbit/urbit.org/tree/master/content/docs.

**Pull request closed without merging**